pkg/jobs/job_info_storage.go

```diff
@@ -0,0 +1,250 @@
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/errors"
+)
+
+// InfoStorage can be used to read and write rows to system.jobs_info table. All
+// operations are scoped under the txn and are are executed on behalf of Job j.
+type InfoStorage struct {
+	j   *Job
+	txn isql.Txn
+}
+
+// InfoStorageWithTxn returns a new InfoStorage with the passed in job and txn.
+func (j *Job) InfoStorageWithTxn(txn isql.Txn) InfoStorage {
+	return InfoStorage{j: j, txn: txn}
+}
+
+// InfoStorageNoTxn returns a new InfoStorage with the passed in job but no txn.
+func (j *Job) InfoStorageNoTxn() InfoStorage {
+	return InfoStorage{j: j}
+}
+
+func (i InfoStorage) checkClaimSession(ctx context.Context) error {
+	row, err := i.txn.QueryRowEx(ctx, "check-claim-session", i.txn.KV(),
+		sessiondata.NodeUserSessionDataOverride,
+		`SELECT claim_session_id FROM system.jobs WHERE id = $1`, i.j.ID())
+	if err != nil {
+		return err
+	}
+
+	if row == nil {
+		return errors.Errorf(
+			"expected session %q for job ID %d but found none", i.j.Session().ID(), i.j.ID())
+	}
+
+	storedSession := []byte(*row[0].(*tree.DBytes))
+	if !bytes.Equal(storedSession, i.j.Session().ID().UnsafeBytes()) {
+		return errors.Errorf(
+			"expected session %q but found %q", i.j.Session().ID(), sqlliveness.SessionID(storedSession))
+	}
+
+	return nil
+}
+
+func (i InfoStorage) get(ctx context.Context, infoKey []byte) ([]byte, bool, error) {
+	if i.txn == nil {
+		var val []byte
+		var exists bool
+		var err error
+		err = i.j.registry.internalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			i.txn = txn
+			val, exists, err = i.get(ctx, infoKey)
+			return err
+		})
+
+		return val, exists, err
+	}
+
+	ctx, sp := tracing.ChildSpan(ctx, "get-job-info")
+	defer sp.Finish()
+
+	j := i.j
+	row, err := i.txn.QueryRowEx(
+		ctx, "job-info-get", i.txn.KV(),
+		sessiondata.NodeUserSessionDataOverride,
+		"SELECT value FROM system.job_info WHERE job_id = $1 AND info_key = $2 ORDER BY written DESC LIMIT 1",
+		j.ID(), infoKey,
+	)
+
+	if err != nil {
+		return nil, false, err
+	}
+
+	if row == nil {
+		return nil, false, nil
+	}
+
+	value, ok := row[0].(*tree.DBytes)
+	if !ok {
+		return nil, false, errors.AssertionFailedf("job info: expected value to be DBytes (was %T)", row[0])
+	}
+
+	return []byte(*value), true, nil
+}
+
+func (i InfoStorage) write(ctx context.Context, infoKey, value []byte) error {
+	if i.txn == nil {
+		return i.j.registry.internalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			i.txn = txn
+			return i.write(ctx, infoKey, value)
+		})
+	}
+
+	ctx, sp := tracing.ChildSpan(ctx, "write-job-info")
+	defer sp.Finish()
+
+	j := i.j
+
+	if j.Session() != nil {
+		if err := i.checkClaimSession(ctx); err != nil {
+			return err
+		}
+	} else {
+		// TODO(during review): Do we want to allow this? CreateAdoptableJobWithTxn
```

Comment:
My vote originally was yes -- you should be allowed  to stage infos for a job that you haven't created yet, let alone adopted yet. 

But I could see drawing a line between creation (just put) versus "update" (delete+put)? i.e. create runs a `select... where info_key` and requires 0 results, whereas update checks the session and runs the blind delete? I donno, as long as I'm allowed to write infos without adopting.

Reply:
short-term suggestion: non-exported helper that takes a checkHasSession bool, exported method pass true except one `WriteWithoutVerifyingSession` which anyone who wants to stage pre-creation can use

Reply:
I added a check that there are no previous rows for this info_key if you don't hold a valid liveness session. I'm not sure if you meant splitting Write itself into Create and Update but I'm leaning to have just Write and have that reject updates without a session.

Reply:
huh so not allowing updates without a session seems to have broken a bunch of tests. I think the reason for most of the breakage is that `LoadJobWithTxn` calls `load` with a job struct that never has the session field set - https://github.com/cockroachdb/cockroach/blob/b407a3aad2d1f37e9ddc956eeb49e840a48533c7/pkg/jobs/registry.go#L877. So `load` is always going to run the branch where we don't have a session set in our in-memory job struct - https://github.com/cockroachdb/cockroach/blob/16d17ff05c9da93b231933bc29a31d4296fb3564/pkg/jobs/jobs.go#L791. Since the loaded job struct that we return never has the session field set any subsequent calls using that loaded job to `Update` will be performed without the session field? So for example `UpdateSchemaChangeJob` always updates the schema change job without a session from what I can tell -https://github.com/cockroachdb/cockroach/blob/2b730a0724e47a1fb70aab0c95c447ca8e295d7c/pkg/sql/schemachanger/scdeps/exec_deps.go#L122.

This breaks our "only allowed to create but not update without a session" invariant. It feels wrong that when we load a job we don't return it with its current liveness session right? We have the `claim_session_id` and `claim_instance_id` that we can unmarshal from `system.jobs` but we'd need to write some method that can fetch us the session object corresponding to those two IDs.

Reply:
I think I'm back to my previous position: if you want to verify you have the session, verify it yourself and don't ask info store reads or writes to verify it for you.

Reply:
The fact that load job doesn't populate the session feels a little wonky to me. Like most callsites Load and then Update so most of our updates are without verifying the correct session. I don't want to try to fix that here though so I'll go back to what I had to begin with.

Reply:
FWIW, we had many bugs where callers (resumers, whatnot) didn't check that they still have the session.  I don't think we should have APIs that will invite such things back. (Though, I admit, I haven't looked closely at this PR, so maybe what I say here is nonesense)

Reply:
We do have some seemingly legitimate call sites (post creation) where you'd want to update the job even if you haven't adopted it ie. have a valid session on it. For example the `ALTER CHANGEFEED` plan hook, import expression walker etc. As David mentioned maybe we want to have two methods:
`Update` and `UpdateWithoutVerifyingSession` and then be deliberate about which one is called from the node that's adopted the job vs from other nodes that don't have a claim on the job but still want to update its state.

This cleanup will probably touch almost all jobs and is a pre-cursor to having two methods in info storage `Write` and `WriteWithoutVerifyingSession`. I agree that this separation would make the API nicer but I think it is outside the scope of our short term goal of migrating payload and progress to jobs_info.

Reply:
> outside the scope
+1